### PR TITLE
add support for buildernames in REST api

### PR DESCRIPTION
**Pull request by tardyp**
Partial Fix for #3407 

## Contributor Checklist:

* [x] I have updated the unit tests
* [x] I have created a file in the `master/buildbot/newsfragment` directory (and read the `README.txt` in that directory)
* [x] I have updated the appropriate documentation

### CODE REVIEW

**Comment by aakashja** [comment]
Can we merge it soon (before next release)?

**Comment by tardyp** [comment]
I need to find some time to finish the unit tests

File: master/docs/developer/database.rst [hunk 1]
```diff
@@ -1457,15 +1457,17 @@ builders
     * ``name``  -- the builder name, a 20-character :ref:`identifier <type-identifier>`
     * ``masterids`` -- the IDs of the masters where this builder is configured (sorted by id)
 
-    .. py:method:: findBuilderId(name)
+    .. py:method:: findBuilderId(name, autoCreate=True)
 
         :param name: name of this builder
         :type name: 20-character :ref:`identifier <type-identifier>`
+        :param autoCreate: create automatically the builder if name not found
```
Comment (seankelly) [inline master/docs/developer/database.rst]: Switch "create" and "automatically".

**Comment by aakashja** [comment]
I tried these commits today. The builder endpoint seems to be working fine for both rest api and UI. However builds endpoint seems to work only for rest api, doesn't work for UI.

e.g.: 
https://nine.buildbot.net/#/builders/buildbot-job/builds/1 would **NOT** work
https://nine.buildbot.net/api/v2/builders/buildbot-job/builds/1 would work

Both of these would work:
https://nine.buildbot.net/api/v2/builders/buildbot-job
https://nine.buildbot.net/#/builders/buildbot-job

The url which doesn't work https://nine.buildbot.net/#/builders/buildbot-job/builds/1, gives 404 error in developer console (screenshot below).

screen shot 2017-07-20 at 10 20 59 pm

In another buildbot instance similar url gives following error:
screen shot 2017-07-20 at 11 20 17 pm png

Seems like some front-end script (which parse buildername from url) might need update as well.

**Comment by tardyp** [comment]
This fix does not handle UI as per comment: https://github.com/buildbot/buildbot/issues/3407#issuecomment-312874963

I already tried it like 6 month ago, and I know it was harder than I though. I can't remember details on why exactly.

**Pull request merged**